File: src/Flagrum.Blender/core/graphics/models/vertexstream.py
from dataclasses import dataclass

from .vertexelement import VertexElement
from ...serialization.messagepack.writer import MessagePackWriter


@dataclass(init=False)
class VertexStream:
    slot: int
    type: int
    stride: int
    offset: int
    elements: list[VertexElement]

    def read(self, reader):
        self.slot = reader.read()
        self.type = reader.read()
        self.stride = reader.read()
        self.offset = reader.read()

        element_count = reader.read()
        self.elements = []
        for i in range(element_count):
            element = VertexElement()
            element.read(reader)
            self.elements.append(element)

    def write(self, writer: MessagePackWriter):
        writer.write(self.slot)
        writer.write(self.type)
        writer.write(self.stride)
        writer.write(self.offset)
        writer.write(self.elements)

File: src/Flagrum.Core/Graphics/Models/VertexStream.cs
using System.Collections.Generic;
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Graphics.Models;

public class VertexStream : IMessagePackItem
{
    public VertexStreamSlot Slot { get; set; }
    public VertexStreamType Type { get; set; }
    public uint Stride { get; set; }
    public uint Offset { get; set; }
    public IList<VertexElement> Elements { get; set; }

    public void Read(MessagePackReader reader)
    {
        Slot = (VertexStreamSlot)reader.Read<int>();
        Type = (VertexStreamType)reader.Read<int>();
        Stride = reader.Read<uint>();
        Offset = reader.Read<uint>();
        Elements = reader.Read<List<VertexElement>>();
    }

    public void Write(MessagePackWriter writer)
    {
        writer.Write((int)Slot);
        writer.Write((int)Type);
        writer.Write(Stride);
        writer.Write(Offset);
        writer.Write(Elements);
    }
}